File: algorithms/code/leetcode/lc129_sum_root_to_leaf_numbers/lc129_sum_root_to_leaf_numbers.py
# https://leetcode.com/problems/sum-root-to-leaf-numbers/

from __future__ import annotations

import unittest


class TreeNode:
    def __init__(
        self,
        val: int = 0,
        left: 'TreeNode | None' = None,
        right: 'TreeNode | None' = None,
    ) -> None:
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def __sumNumbers(self, root: TreeNode | None, prefix: int) -> int:
        if root is None:
            return 0
        prefix = prefix * 10 + root.val
        if root.left is None and root.right is None:
            return prefix
        left = self.__sumNumbers(root.left, prefix)
        right = self.__sumNumbers(root.right, prefix)
        return left + right

    def sumNumbers(self, root: TreeNode | None) -> int:
        return self.__sumNumbers(root, 0)


class TestCode(unittest.TestCase):
    def test_example(self) -> None:
        root = TreeNode(1)
        root.left = TreeNode(2)
        root.right = TreeNode(3)
        self.assertEqual(25, Solution().sumNumbers(root))

    def test_nothing(self) -> None:
        self.assertEqual(0, Solution().sumNumbers(None))

    def test_left(self) -> None:
        root = TreeNode(1)
        root.left = TreeNode(2)
        self.assertEqual(12, Solution().sumNumbers(root))

    def test_right(self) -> None:
        root = TreeNode(1)
        root.right = TreeNode(3)
        self.assertEqual(13, Solution().sumNumbers(root))

File: algorithms/code/leetcode/lc129_sum_root_to_leaf_numbers/LC129SumRootToLeafNumbers.java
package leetcode.lc129_sum_root_to_leaf_numbers;

/**
 * https://leetcode.com/problems/sum-root-to-leaf-numbers/
 * #medium
 */
public final class LC129SumRootToLeafNumbers {
    public static class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;

        TreeNode(final int x) {
            val = x;
        }
    }

    private int sumNumbers(final TreeNode root, final int prefixValue) {
        int prefix = prefixValue;
        if (root == null) {
            return 0;
        }
        prefix = prefix * 10 + root.val;
        if (root.left == null && root.right == null) {
            return prefix;
        }
        return sumNumbers(root.left, prefix) + sumNumbers(root.right, prefix);
    }

    public int sumNumbers(final TreeNode root) {
        return sumNumbers(root, 0);
    }
}
package leetcode.lc129_sum_root_to_leaf_numbers;

import org.junit.jupiter.api.Test;

import leetcode.lc129_sum_root_to_leaf_numbers.LC129SumRootToLeafNumbers.TreeNode;

import static org.junit.jupiter.api.Assertions.assertEquals;

public final class LC129SumRootToLeafNumbersTests {
    @Test
    public void testExample() throws Exception {
        TreeNode root = new TreeNode(1);
        root.left = new TreeNode(2);
        root.right = new TreeNode(3);
        assertEquals(25, new LC129SumRootToLeafNumbers().sumNumbers(root));
    }

    @Test
    public void testNothing() throws Exception {
        assertEquals(0, new LC129SumRootToLeafNumbers().sumNumbers(null));
    }

    @Test
    public void testLeft() throws Exception {
        TreeNode root = new TreeNode(1);
        root.left = new TreeNode(2);
        assertEquals(12, new LC129SumRootToLeafNumbers().sumNumbers(root));
    }

    @Test
    public void testRight() throws Exception {
        TreeNode root = new TreeNode(1);
        root.right = new TreeNode(3);
        assertEquals(13, new LC129SumRootToLeafNumbers().sumNumbers(root));
    }
}
